packages/firestore/src/remote/datastore.ts

```diff
@@ -103,6 +106,10 @@ class DatastoreImpl extends Datastore {
       this.appCheckCredentials.getToken()
     ])
       .then(([authToken, appCheckToken]) => {
+        const path = toResourcePath(databaseId, resourcePath)
+          .toArray()
+          .map(encodeURIComponent)
```

Comment:
I think this entire PR looks really great and I'm going to LGTM it. My only reservation is that the path component encoding should really be a property of the `Connection` rather than baked into the Datastore. The path encoding is a requirement of `RestConnection` that is bleeding into `Datastore`. 

One solution could be to modify the `Connection` interface to change the `path` argument from `string` to `string[]`. Then, GrpcConnection ignores the path anyways so there would be nothing to do there. However, in RestConnection it could to the `encodeURIComponent` and `join('/')` in its `invokeRPC()` and `invokeStreamingRPC()` functions. If you agree with a refactoring like this, that's totally fine to do in a follow-up PR as a "clean up".

Reply:
That makes sense. I made this change now because it would likely never get prioritized in the future, but I'm happy with the outcome.